modules/gapi/src/executor/gthreadedexecutor.cpp

```diff
@@ -0,0 +1,525 @@
+    }
+public:
+    Output(cv::gimpl::Mag &m, const std::vector<RcDesc> &rcs, std::mutex &mutex)
+        : mag(m), m_mutex(mutex)
+    {
+        set(rcs);
+    }
+
+    void verify()
+    {
+        if (eptr)
+        {
+            std::rethrow_exception(eptr);
+        }
+    }
+};
+
+
+void cv::gimpl::GThreadedExecutor::initResource(const ade::NodeHandle &nh, const ade::NodeHandle &orig_nh) {
+    const Data &d = m_gm.metadata(orig_nh).get<Data>();
+
+    if (   d.storage != Data::Storage::INTERNAL
+        && d.storage != Data::Storage::CONST_VAL)
+        return;
+
+    // INTERNALS+CONST only! no need to allocate/reset output objects
+    // to as it is bound externally (e.g. already in the m_res)
+
+    switch (d.shape)
+    {
+    case GShape::GMAT:
+        {
+            // Let island allocate it's outputs if it can,
+            // allocate cv::Mat and wrap it with RMat otherwise
+            GAPI_Assert(!nh->inNodes().empty());
+            const auto desc = util::get<cv::GMatDesc>(d.meta);
+            auto& exec = m_gim.metadata(nh->inNodes().front()).get<IslandExec>().object;
+            auto& rmat = m_res.slot<cv::RMat>()[d.rc];
+            if (exec->allocatesOutputs()) {
+                rmat = exec->allocate(desc);
+            } else {
+                Mat mat;
+                createMat(desc, mat);
+                rmat = make_rmat<RMatOnMat>(mat);
+            }
+        }
+        break;
+
+    case GShape::GSCALAR:
+        if (d.storage == Data::Storage::CONST_VAL)
+        {
+            auto rc = RcDesc{d.rc, d.shape, d.ctor};
+            magazine::bindInArg(m_res, rc, m_gm.metadata(orig_nh).get<ConstValue>().arg);
+        }
+        break;
+
+    case GShape::GARRAY:
+        if (d.storage == Data::Storage::CONST_VAL)
+        {
+            auto rc = RcDesc{d.rc, d.shape, d.ctor};
+            magazine::bindInArg(m_res, rc, m_gm.metadata(orig_nh).get<ConstValue>().arg);
+        }
+        break;
+    case GShape::GOPAQUE:
+        // Constructed on Reset, do nothing here
+        break;
+    case GShape::GFRAME: {
+        // Should be defined by backend, do nothing here
+        break;
+    }
+    default:
+        GAPI_Error("InternalError");
+    }
+}
+
+cv::gimpl::IslandActor::IslandActor(const std::vector<RcDesc>          &in_objects,
+                                    const std::vector<RcDesc>          &out_objects,
+                                    std::shared_ptr<GIslandExecutable> isl_exec,
+                                    Mag                                &res,
+                                    std::mutex                         &m)
+    : m_in_objs(in_objects),
+      m_out_objs(out_objects),
+      m_isl_exec(isl_exec),
+      m_res(res),
+      m_mutex(m) {
+}
+
+void cv::gimpl::IslandActor::run() {
+    cv::gimpl::GThreadedExecutor::Input  i{m_res,  m_in_objs,  m_mutex};
+    cv::gimpl::GThreadedExecutor::Output o{m_res,  m_out_objs, m_mutex};
```

Comment:
Why are the inputs/outputs created right before run? I believe they could be created in the ctor and then reused across runs? They are lazy and their sense and only do the job when are called to?

Reply:
> Why are the inputs/outputs created right before run? I believe they could be created in the ctor and then reused across runs? They are lazy and their sense and only do the job when are called to?

Mostly because they are defined in `*.cpp` file, we do the same for the regular `GExecutor`.
Again, no problems, to address it in this PR :)